### PR TITLE
Add ROCm support for Jax Triton

### PR DESCRIPTION
**Pull request by rahulbatra85**

### CODE REVIEW

**Comment by rahulbatra85** [comment]
@hawkinsp  @sharadmv 
Hi, this PR adds support for Jax-Triton API for ROCm. Please provide feedback. Thanks!

File: jaxlib/gpu/triton_kernels.cc [hunk 1]
```diff
@@ -25,22 +25,24 @@
...
 
-#define CUDA_RETURN_IF_ERROR(expr) JAX_RETURN_IF_ERROR(JAX_AS_STATUS(expr))
+#define GPU_RETURN_IF_ERROR(expr) JAX_RETURN_IF_ERROR(JAX_AS_STATUS(expr))
 
 
 namespace jax::JAX_GPU_NAMESPACE {
 namespace {
 
-constexpr uint32_t kNumThreadsPerWarp = 32;
+//constexpr uint32_t kNumThreadsPerWarp = 32;
```
Comment (hawkinsp) [inline jaxlib/gpu/triton_kernels.cc:36]: Is this change intentional?

File: jaxlib/rocm/BUILD [hunk 2]
```diff
@@ -286,6 +286,67 @@ pybind_extension(
     ],
 )
```
Comment (hawkinsp) [inline jaxlib/rocm/BUILD:288]: Please keep this file named BUILD.bazel.

(The distinction is mostly relevant to Google's internal build system, which ignores BUILD.bazel but does look at BUILD.)

File: jaxlib/gpu_triton.py [hunk 3]
```diff
@@ -15,17 +15,33 @@
 from jaxlib import xla_client
```
Comment (hawkinsp) [inline jaxlib/gpu_triton.py:16]: This is suboptimal, because in general I'm trying to maintain the property that CUDA and ROCM could be simultaneously imported in the same JAX program. The code below forces you to pick one: if the ROCM code is present, it will override the CUDA code. Could we avoid that?

(I mean, I admit it's unlikely people do this often, but nothing physically stops you from having a machine with both kinds of GPU, I think?)

Under the upcoming PJRT plugin work, it should be fairly easy to install plugins for both GPUs into JAX at the same time, as well.

It's not absolutely mandatory that we do this, but is there a way we could?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
This looks pretty good

File: jaxlib/gpu_triton.py (hunk 3 above)
Comment (hawkinsp) [reply jaxlib/gpu_triton.py:16]: This comment wasn't addressed.

**Review by hawkinsp — COMMENTED** [review]
Looks fine overall. Can you at least look at the gpu_triton.py comment? Would that be hard to do?

**Comment by rahulbatra85** [comment]
https://github.com/google/jax/pull/18078#discussion_r1362405509ks fine overall. Can you at least look at the gpu_triton.py comment? Would that be hard to do?

@hawkinsp I wrote this above. Copy/pasting it here.

Thanks for your feedback!

I see your point, but I will need to think how to best do this.

We are doing something similar for other API like blas, solver, sparse. I guess those will also need to be changed. Right? Also, do you think any changes will be needed to the Triton API itself? Like exposing the fact that multiple platforms are available and when TritonKernelCall is made, and then another argument can be passed to specify which platform to execute the kernel on. May be also need a notion of default platform.

For now, can we merge this PR and then address this separately?

Thanks!

**Comment by hawkinsp** [comment]
> [#18078 (comment)](https://github.com/google/jax/pull/18078#discussion_r1362405509ks) fine overall. Can you at least look at the gpu_triton.py comment? Would that be hard to do?
> 
> @hawkinsp I wrote this above. Copy/pasting it here.
> 
> Thanks for your feedback!
> 
> I see your point, but I will need to think how to best do this.
> 
> We are doing something similar for other API like blas, solver, sparse. I guess those will also need to be changed. Right?

Actually, no, I don't think we are. The CUDA-specific stuff ends up in a directory named `jaxlib/cuda`, and should in no way conflict with a parallel installation of the ROCM stuff (although that's admittedly very hard to do at the moment).

It will become *considerably* easier to do this soon, because we're going to split all CUDA-specific stuff into a separate plugin package, named something like `jax-cuda-...`. There will no longer be GPU and non-GPU jaxlib variants, only a common `jaxlib` with a CUDA plugin package. And presumably you'll want to ship the ROCM code the same way.

> Also, do you think any changes will be needed to the Triton API itself? Like exposing the fact that multiple platforms are available and when TritonKernelCall is made, and then another argument can be passed to specify which platform to execute the kernel on. May be also need a notion of default platform.

I actually suspect not eventually. Essentially we want to ship a copy of triton with the plugin, so you can have two different copies of triton, one with each plugin. (In fact, we *do* ship a copy of triton with XLA, we're just not using that copy from JAX yet.)

 > For now, can we merge this PR and then address this separately?

Sure.

**Review by hawkinsp — APPROVED** [review]

**Comment by hawkinsp** [comment]
There's a lint failure.

**Comment by hawkinsp** [comment]
(The lint failure is still present.)

**Comment by rahulbatra85** [comment]
> (The lint failure is still present.)

Sorry, about that. 

I pushed another change. I think it should be fixed now

**Review by hawkinsp — APPROVED** [review]

**Comment by rahulbatra85** [comment]
ah GPU build failed now.

File: jaxlib/gpu_triton.py (hunk 3 above)
Comment (i-chaochen) [reply jaxlib/gpu_triton.py:16]: I think you're right @hawkinsp, if xla can allow multiple stream executor co-exist, we should be capable of achieving it. As a side note, I wonder is TPU and NV-GPUs can coexist in a machine?

**Review by i-chaochen — COMMENTED** [review]
I think you're right @hawkinsp, if xla can allow multiple stream executor co-exist, we should be capable of achieving it. As a side note, I wonder is TPU and NV-GPUs can coexist in a machine?
